Fix offset issue in DELTA_BINARY_PACKED decoding

sadikovi opened the pull request:
This PR fixes issue when reading values from DELTA_BINARY_PACKED decoder, we would only read values we need and load deltas that we need. This poses a problem when we want to use this decoder in other decodings like DELTA_LENGTH_BYTE_ARRAY and DELTA_BYTE_ARRAY.

1. The main problem is that DELTA_BINARY_PACKED decoder only reads bytes required to reconstruct requested values, it will not load full mini block. This causes offset problem when using this decoder in other decoders like DELTA_LENGTH_BYTE_ARRAY and DELTA_BYTE_ARRAY.

The reason it causes the problem is as follows:
- Let's say we have a vector of bytes like this: `[128, 1, 4, 3, 58, 28, 6, 0, 0, 0, 0, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`, with trailing zeros. This set of bytes is encoded vector of 3 values: `[29, 43, 89]`. 
- When we want to decode this, we set this as input data and ask for 3 values. 
- When we read all 3 values, there will be some values left in mini block (31) that we have not touched, because we only need 3 values. 
- The problem is that bit reader does not advance read offset, until we actually read those bytes, resulting in shorter offset: before fix it would return 11, actually should return 34.

This is constantly reproducible when running delta length byte array encoding on vector of larger than 1 value. I added unit test that covers this issue.

2. Another issue that I found was that byte offset that is returned by `BitReader` reports `+1` bytes. This should be just `byte_offset + ceil(bit_offset, 8)` to cover case when bit offset is 0.

3. Also removed assertion check in `init_block` and test. This is internal method, no need to check, because we invoke it only when bit reader is initialized. I assume having assertion there has some performance overhead.

sadikovi commented:
@sunchao could you review this PR, please? 

This should help with adding other encodings/decodings that use this decoder. I looked at C++ implementation and did not find how they were doing this (or maybe it has a bug when decoding byte array values). So mainly I used parquet-mr implementation for the reference.

parquet-mr implementation load values eagerly (all in buffer). I decided to do something in between and load deltas eagerly for a mini block, every time new mini block is read, we load deltas. This fixes the offset problem, since it was coming from lazily loading of data from bit reader.

Would like to know your view and suggestions on this? Should we do this similar to parquet-mr or parquet-cpp? Thanks.

sadikovi commented:
Actually, I am still testing it with delta byte array encoding.

sunchao commented:
Thanks @sadikovi ! I'll take a look soon (probably this weekend)

sadikovi commented:
Tested this fix with delta byte array encodings patch - everything works as expected, now delta bit pack decoder reports all necessary bytes in offset, so byte arrays' suffixes are read correctly.

sadikovi commented:
I will probably spend some time afterwards on improving delta bit pack encoding performance.

sadikovi commented on src/util/bit_util.rs:
```diff
@@ -392,7 +392,7 @@ impl BitReader {
   /// Gets the current byte offset
   #[inline]
   pub fn get_byte_offset(&self) -> usize {
-    self.byte_offset + self.bit_offset / 8 + 1
```
I am a bit confused here. Is `bit_offset` within `byte_offset`? My original thinking was that byte_offset + bit_offset/8 + 1 will yield +2 bytes occasionally for a case when bit_offset is within byte_offset.

My tests pass when I remove `+1`, and I remember comparing it to parquet-mr reads, but I may be wrong here. Anyway, will review this.

sunchao replied:
bit_offset is the offset with in `buffered_values`, which is a 64-bit value. The `buffered_values` is positioned after `byte_offset`.

sadikovi commented:
@sunchao Thanks for the review. The problem with current implementation of DeltaBitPackDecoder (encoder is correct) is that when reading values we materialize them "lazily", we do not need all values, store them in buffer and copy them when `get()` method is called.

This has implications to decodings that use delta bit packing. In DELTA_LENGTH_BYTE_ARRAY decoder we store `lengths` + `data`. In order to know where `lengths` end decoder needs to read all values (this is how parquet-mr implementation works). Otherwise, when you call offset, it returns position that is part of delta bit packing data, and results in wrong values when decoding. For example,

Encoded data:
```
[128, 0, 4, 8, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8]
  ^------------------------------^
   original delta bit pack data (what we are supposed to read)

  ^---------^
   actual data (where byte_offset points to in bit reader, after decoding necessary values)
```

As DELTA_BYTE_ARRAY decoder uses DELTA_LENGTH_BYTE_ARRAY it has the same problems, actually 2x, because it uses to encode lengths as well as delta length byte array decoding.

I have several questions:
- How does DELTA_LENGTH_BYTE_ARRAY and DELTA_BYTE_ARRAY work in parquet-cpp? How do they know where delta bit packing data ends and actual data begins (since we do not store any offsets)
- My solution might also be incorrect, but my assumption is based on the fact that we would decode all lengths, and the last mini block should be read fully. Is this a good approach?

Again, it all comes down to how parquet-cpp handles this, and so far, I could not find the place where they were using similar code to separate `lengths` part from `data` part in DELTA_LENGTH_BYTE_ARRAY decoding.

I will have a look at this again, since I have some doubts about implementation.

sadikovi commented:
I also patched the code in those decodings, so I will open PR with those, after we have sorted out delta bit packing decoding! I just need some help with understanding how parquet-cpp knows how to split lengths from actual data in DELTA_LENGTH_BYTE_ARRAY decoding:)

sunchao commented:
@sadikovi Got it. Thanks for the explanation. I think parquet-cpp doesn't follow the Parquet format standard for DELTA_LENGTH_BYTE_ARRAY and DELTA_BYTE_ARRAY. See [here](https://github.com/apache/parquet-cpp/blob/master/src/parquet/encoding-internal.h#L883) for example:

parquet-cpp for DELTA_LENGTH_BYTE_ARRAY assumes that the input data is in the following format:

```
- length of delta_bit_packed data for the byte array lengths (encoded in 4 bytes)
- delta_bit_packed data for the byte array lengths
- data for the byte array data
```

The first part is not in the Parquet format.

So yes, I think it makes sense to follow the parquet-mr approach.

sadikovi commented on src/util/bit_util.rs:
```diff
@@ -392,7 +393,7 @@ impl BitReader {
   #[inline]
   pub fn get_byte_offset(&self) -> usize {
-    self.byte_offset + self.bit_offset / 8 + 1
+    self.byte_offset + ceil(self.bit_offset as i64, 8) as usize
```
@sunchao you are right, we should use ceil here instead of `+1`, this covers case when bit offset is 0. I added tests showing the actual offset for each value we get.

sadikovi commented:
@sunchao I updated the code, could you review again? I updated the ASCII chart indicating the start of buffer, because I was reading it from left to right.

Is there a conversion overhead from using `ceil(offset as i64, 8) as usize`? Would it be better to write it for usize specifically?

Thanks.

sunchao commented:
> Is there a conversion overhead from using ceil(offset as i64, 8) as usize? Would it be better to write it for usize specifically?

It seems this will be optimized away by the compiler. Check [here](https://godbolt.org/g/aJZAYU) for the generated assembly. 

> Do you agree with overall approach? I do it similar to parquet-mr, but parquet-mr decodes all values eagerly, we do it only for current mini block deltas. Should we be doing it as parquet-mr? It actually does not cause any issues, apart from the fact that we need full mini block for when this decoder is used as part of other decoders.

Yes I think it should sufficient to just fully read the current mini block. In DELTA_LENGTH_BYTE_ARRAY and DELTA_BYTE_ARRAY we would fully read all the values for lengths at the beginning, so there wouldn't be any mini block left to read after. The only potential issue is the padding in a single mini block, which is addressed by this commit.

sadikovi commented:
Okay, thanks for the clarifications! Do you want me to make any other changes to the code? Thanks!

sunchao commented:
@sadikovi PR looks good. I've merged it. Thanks!

sadikovi commented:
Thanks for merging!
